### PR TITLE
Show confirmation when leaving multistore form without save

### LINKED ISSUE #23741: Multistore: warning before changing context when the form has been modified
Following the [multistore header specifications](https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/multistoregeneralspecs.md):

> If the user tries to change context without having saved, a modal appears: 
Title: "Switch context?"
Body: "Switching context now will discard all unsaved changes." 
2 CTA: "Switch context anyway" and "Cancel" 
If the user clicks on "Switch context anyway", then modifications are not saved and the context is changed for the one selected. 
If the user clicks on "Cancel", then the context is not changed.

### PR DESCRIPTION
**Pull request by sowbiba**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Show confirmation when leaving multistore form without saving. If any modifiction is done, when you quit the page, without clicking in the submit button, the confirm modal will show up
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #23741
| How to test?      | Enable multishop, create a second shop, Go to maintenance page, edit a field and try to change page or context. See the confirm bow shown
| Possible impacts? | -

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25706)

### CODE REVIEW

**Review by Progi1984 — COMMENTED** [review]
@sowbiba Thank you for your contribution. Could you fix ESLint errors ?

**Comment by Progi1984** [comment]
@PrestaShop/product-team I added the label "Waiting for PM" as the issue #23741 is waiting for PM.

**Review by matthieu-rolland — COMMENTED** [review]
LGTM for me, waiting for PM feedback before aproval

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/components/multistore-form.ts]: I think this string needs to be translated into multiple langs

Add the translation from PHP directly inside a JS object, or add it in the markup in a data-attribute and retrieve it there!

Comment (sowbiba) [reply admin-dev/themes/new-theme/js/components/multistore-form.ts]: Just seen your comment. This message is not shown by browsers (see the comment above the line). I put a random message

**Comment by marionf** [comment]
I can't test this PR because of this issue: https://github.com/PrestaShop/PrestaShop/issues/25878
I will take a look when this issue will be fixed

**Comment by marionf** [comment]
@sowbiba 

The default modal of the web browser is really confusing because it's asking if you want to leave the website while you just want to switch context 😅 
If I check a box, uncheck it just after and try to change context, the modal is displayed while it shouldn't because nothing has changed.

If we can't customize the content of the modal and display it only in the case of context switching (not page), I think it's better to do nothing because the web browser modal brings too much confusion in the user experiance

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (kpodemski) [inline admin-dev/themes/new-theme/js/components/multistore-form.ts]: @sowbiba I think you could bind to the context selector switcher and show a dedicated modal with a proper message

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (PierreRambaud) [reply admin-dev/themes/new-theme/js/components/multistore-form.ts]: It's a fallback message, so it needs to use a translated string

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/components/multistore-form.ts]: If the purpose is to avoid changing context when the user switch shop context then as @kpodemski suggested you can trigger a confirm modal based on the multistore header action rather than the `onbeforeunload` event.

The advantage is that you can use a custom confirm modal, it is of course less generic, but it actually matches more the behaviour around multi store. Because if register on unload event even clicking a link in the menu will trigger the blocking warning, and that's not what we aim for if I got it right.

If this more detailed check and custom modal is accepted then you can indeed add the translated message in a data attribute on the form (in the extension just like you didi for the extra class).

**Review by jolelievre — COMMENTED** [review]
@sowbiba I think the modal can be improved (nice confirm modal, and translated message). But probably only I you bond it to the multistore header instead of the generic unload event.

That's to be confirmed with product team, but it would be a more logical behaviour IMHO. We want to prevent a shop context switch not any exit from the page right? (like clinking in the menu).

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/components/multistore-form.ts]: Besides if we can't use our custom message then the user won't understand why we are blocking the change. The message should be about switching shop context no?

**Review by matks — COMMENTED** [review]
@jolelievre raises an important point 👍  I agree it seems like a nice improvement

**Comment by marionf** [comment]
@sowbiba 

If I check a box, uncheck it just after without changing anything on the page and try to change context, the modal is displayed while it shouldn't because nothing has been changed on the page.

The content of the modal is currently "Changing context without saving your modifications will make you loose them. Do you want to quit ?"
It should be "You are about to switch context. Do you want to save your modifications before?"

The 2 CTA are currently "No, stay" and "Yes, quit"
It should be "Save" and "Cancel"
If the user clicks on "Save", then modifications are saved and the context is changed for the one selected.
If the user clicks on "Cancel", then modifications are not saved and the context is changed for the one selected.

**Comment by sowbiba** [comment]
> @sowbiba
> 
> If I check a box, uncheck it just after without changing anything on the page and try to change context, the modal is displayed while it shouldn't because nothing has been changed on the page.
> 
> The content of the modal is currently "Changing context without saving your modifications will make you loose them. Do you want to quit ?" It should be "You are about to switch context. Do you want to save your modifications before?"
> 
> The 2 CTA are currently "No, stay" and "Yes, quit" It should be "Save" and "Cancel" If the user clicks on "Save", then modifications are saved and the context is changed for the one selected. If the user clicks on "Cancel", then modifications are not saved and the context is changed for the one selected.

The behaviour for the buttons you asked doesn't offer the possibility for the user to review changes before leave

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (Julievrz) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]:
```suggestion
                'modal.confirm_leave.title': 'Switch context?'|trans({}, 'Admin.Notifications.Warning'),
```

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (Julievrz) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]:
```suggestion
                'modal.confirm_leave.body': 'Switching context now will discard all unsaved changes.'|trans({}, 'Admin.Actions'),
```

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (Julievrz) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]:
```suggestion
                'modal.confirm_leave.cancel': 'Cancel'|trans({}, 'Admin.Actions'),
```

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (Julievrz) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]:
```suggestion
                'modal.confirm_leave.confirm': 'Switch context without saving'|trans({}, 'Admin.Actions')
```

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (marionf) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]:
```suggestion
                'modal.confirm_leave.confirm': 'Switch context anyway'|trans({}, 'Admin.Actions')
```

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (Julievrz) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]:
```suggestion
                'modal.confirm_leave.title': 'Switch context?'|trans({}, 'Admin.Notifications.Warning'),
```

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (Julievrz) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]:
```suggestion
                'modal.confirm_leave.body': 'Switching context now will discard all unsaved changes.'|trans({}, 'Admin.Notifications.Warning'),
```

**Review by Julievrz — COMMENTED** [review]
Sorry, I had missed the domains. Thank you in advance.

File: admin-dev/themes/new-theme/js/components/modal/confirm-modal.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/components/modal/confirm-modal.ts]: Why do you need this? This function matches the `closeCallback` from the parent modal component which doesn't use arguments:
https://github.com/PrestaShop/PrestaShop/blob/c71f1a345c146bb546651e4a00c69b4407b9501f/admin-dev/themes/new-theme/js/components/modal/modal.ts#L197-L199

File: src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php:90]: Since this extension is bound to `MultistoreConfigurationType` it means it should only add the class in the configuration forms right?

So it means, this way, it shouldn't work on other forms like entities (products, categories, ...). Is it the expected behaviour to only handle the configuration forms? If so, then where did you have your bug with tinymce? I thought it was related to product page, does it mean there are configuration forms that use Tinymce?

Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php:90]: Also the addition of this class should be conditioned to the multistore feature, when the feature is disabled the class shouldn't be added

**Review by jolelievre — COMMENTED** [review]
Thanks @sowbiba 
I have a few questions ^^

File: admin-dev/themes/new-theme/js/components/modal/confirm-modal.ts
Comment (sowbiba) [reply admin-dev/themes/new-theme/js/components/modal/confirm-modal.ts]: I needed it when I wanted to "save when user cancels" the confirm modal. But I don't need it anymore. Forgot to revert

File: src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php
Comment (sowbiba) [reply src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php:90]: If I understood well, any form type multistore compatible must have MultistoreConfigurationType as parent, like we have it in MaintenanceType @matthieu-rolland 

The bug with tinyMCE is when the "multistore compatible type" has textarea fields

Comment (kpodemski) [reply src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php:90]: @jolelievre I think it's because multi-store is always there, but sometimes you have 1 store, and sometimes more?

**Review by kpodemski — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php]:
```suggestion
        $resolver->setNormalizer('attr', function (Options $options, array $value): array {
```

**Review by Progi1984 — DISMISSED** [review]
@sowbiba I approve but if you can check the small typehint in my comment.

**Comment by marionf** [comment]
@sowbiba I no longer see the modal at all

**Comment by sowbiba** [comment]
> @sowbiba I no longer see the modal at all

Saw it together. Just have to build assets.
@marionf still have issues with checkboxes. I will try to reproduce it

**Review by matks — COMMENTED** [review]
Current code looks OK to me 👍  waiting for latest issues with TinyMCE to be solved before approve

**Comment by matks** [comment]
@sowbiba ready for review or WIP?

**Review by matks — DISMISSED** [review]

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (kpodemski) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig:29]: @sowbiba 

Wouldn't it be better to follow the same style for translations as in other views? Example:

```
{% set js_translatable = {
  "Customer search - group default": 'default'|trans({}, 'Admin.Orderscustomers.Feature'),
  "Customer search - group label single": 'Group'|trans({}, 'Admin.Orderscustomers.Feature'),
  "Customer search - group label multiple": 'Groups'|trans({}, 'Admin.Orderscustomers.Feature'),
}|merge(js_translatable)
%}
```

`window.translate_javascripts['Customer search - group label multiple']`

**Review by atomiix — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php]:
```suggestion
            $value['class'] = 'js-multishop-form-row';
            if (!empty($value['class'])) {
                $value['class'] .= ' ' . $value['class'];
            }
```

**Review by Progi1984 — COMMENTED** [review]
A small feedback

File: admin-dev/themes/new-theme/js/components/components-map.ts
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/components/components-map.ts]:
```suggestion
    header: '.js-header-multishop',
```

Prefix every js-classes so we don't use these to style and we can remove/refacto them without thinking

File: admin-dev/themes/new-theme/js/components/components-map.ts
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/components/components-map.ts]:
```suggestion
    item: '.js-multishop-modal-item',
```

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/components/multistore-form.ts]:
```suggestion
          .replace(/<[^>]*>?/gm, '') // remove html tags
          .replace(/\r?\n|\r/g, '') //remove line breaks
          .replace(/(\.\s)/g, '.'); //remove space after dots added by TinyMCE
```

If you do this, it means that if I modify some HTML tags inside tinyMCE, I can quit the page without having the confirmation modal, which is wrong

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/components/multistore-form.ts]: I also have doubts regarding his cleaning of TinyMCE's content 🤔. Comparing the text content only can make sense in other use cases but here the HTML content should also be considered

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]: I don't think we should rely on this old ugly js global object to handle translations of a component In all our recent components we integrate translated content via data attributes

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/components/multistore-form.ts]: What if the `<a>` comes from a module, doesn't have an href because it's wrong and should use a button?

What about buttons inside the header? In case a module adds a button + some JS

```suggestion
          if (targetUrl && typeof targetUrl === 'string') {
            showConfirmModal(targetUrl);
          }
```

So you use typescript types

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/components/multistore-form.ts]: I'm not sure this is the right way to init your multishop form. Because it means if you don't have a tinymce the `tinymceInitialized` event is never triggered and so the multistore form is never initialized either

File: admin-dev/themes/new-theme/js/components/tinymce-editor.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/components/tinymce-editor.js]: The event emitter is used differently somewhere else in this file by doing `EventEmitter.emit` This is not consistent, either use the same convention as the one previously present or update it all

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/components/tinymce-editor.js]:
```suggestion
    window.prestashop.instance.eventEmitter.emit('tinymceInitialized', {});
```

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/components/tinymce-editor.js]: If you choose this convention you should use the instance not the component Even if it's the same for this specific component it's important to keep a consistent use everywhere

File: admin-dev/themes/new-theme/js/types/global.d.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/types/global.d.ts]: There might be a dedicated type for TinyMCE that could be used and more precise Maybe @NeOMakinG you know about some types that may be imported and used?

File: src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php]: Hmm no @Progi1984  your suggestion is removing the initial value on first line so you can't use it after But it could be simplified indeed

Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Extension/MultistoreConfigurationTypeExtension.php]:
```suggestion
            $value['class'] = trim('js-multishop-form-row ' . ($value['class'] ?? ''));
```

File: src/PrestaShopBundle/Form/Admin/Type/TranslateType.php
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Type/TranslateType.php:98]: You should comment this part that was not so easy to find out

Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Type/TranslateType.php:98]:
```suggestion
            // This form type is a composition of other sub types, so the disabled option should be transmitted to children, especially when dealing with rich formatted text inputs
            if (isset($options['disabled'])) {
```

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (jolelievre) [reply src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]: This could be bound to the multistore modal component This would also allow to init your component at all time and not only when tinymce is used

File: src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php
Comment (jolelievre) [inline src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php:165]: This part took us a few hours to figure out you need to comment it in details

Comment (jolelievre) [reply src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php:165]:
```suggestion
        // The updateCurrentField is called multiple times in the form lifecycle, once during init on ORE_SET_DATA event and a second time when the form is submitted on PRE_SUBMIT event
        // Both times the form content needs to be adapted and the rules are different for both cases, on submit the data from the requests overrides the initial data from the configuration
        $fieldName = $childElement->getName();
```

File: src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php
Comment (jolelievre) [inline src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php:168]:
```suggestion
        // When the checkbox has been manually disabled in the form data the input cannot be disabled
        if ($form->has($enablerFieldName) && !empty($data[$enablerFieldName])) {
```

File: src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php
Comment (jolelievre) [inline src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php:173]:
```suggestion
            // When the form is not submitted or no data is found in the request the default state is defined by the configuration
            $isDisabled = !$this->multiStoreContext->isAllShopContext() && !$isOverriddenInCurrentContext;
```

File: src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php
Comment (jolelievre) [inline src/PrestaShopBundle/Service/Form/MultistoreCheckboxEnabler.php:191]: @matthieu-rolland we removed this field because we realized it was actually never used So its presence was confusing

File: admin-dev/themes/new-theme/js/components/tinymce-editor.js
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/components/tinymce-editor.js]: By the way are you sure you need this event, since there is already a `tinymceEditorSetup` event triggered during the editor setup

**Review by jolelievre — CHANGES_REQUESTED** [review]
Thanks @sowbiba 

Most of the PR seems relevant I just have a few doubts on some parts, and I think it's critical to add some comments on complex parts AND update your PR description to justify all the changes that were made in this PR

Especially explain why the submit event is now watched, why the form must adapt to data and all this Because it is way out of the scope of the initial issue/PR so it needs to be documented to understand and review this PR correctly

File: admin-dev/themes/new-theme/js/types/global.d.ts
Comment (NeOMakinG) [reply admin-dev/themes/new-theme/js/types/global.d.ts]: https://www.npmjs.com/package/@types/tinymce

Use the same version as we use in the project

Comment (NeOMakinG) [reply admin-dev/themes/new-theme/js/types/global.d.ts:26]: https://www.npmjs.com/package/@types/tinymce

Use the same version as we use in the project

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (sowbiba) [reply admin-dev/themes/new-theme/js/components/multistore-form.ts]: On page load, the values in textarea is different of the TinyMCE content. The value in the textarea is the raw value, Tinymce enrich the original value by adding HTML tags and formatting like a space after the dots

That's why I clean them all before comparing values

File: src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig
Comment (sowbiba) [reply src/PrestaShopBundle/Resources/views/Admin/Multistore/header.html.twig]: This change was asked in the review process. Maybe the convention is not known enough

File: admin-dev/themes/new-theme/js/components/multistore-form.ts
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/components/multistore-form.ts]: But when you save then all the additional content added by TinyMCE is indeed saved in the DB no?
So it's only the initial data from the fixtures which don't match and after a save I believe it is the same?

Besides with this modif I still don't see how you can see any difference when only the HTML tags are modified 🤔

**Review by Julievrz — DISMISSED** [review]
Wording ✔️

**Review by NeOMakinG — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]
Thanks @sowbiba

**Review by PierreRambaud — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @sowbiba,

Could you please check the Conflicting files?

Thank you!

**Review by NeOMakinG — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]
Thank you @sowbiba

**Review by khouloudbelguith — CHANGES_REQUESTED** [review]
Hi @sowbiba,

I have some issues following these steps:

**Issue1**
1. Enable Multistore & Create a new shop
2. Go to BO > Shop Parameters > Maintenance page & select a specific shop
3. Make a change & click out on another page 
4. See error: the Modal is not displayed

I attached a screen record

https://user-images.githubusercontent.com/16067358/154268281-423c3700-d889-4bf4-a3aa-e0b63fb66ef5.mp4

**Issue2**
1. Enable Multistore & Create a new shop
2. Go to BO > Shop Parameters > Maintenance page & select a specific shop
3. Make a change & don't save
4. Click on the header, search a shop & choose it
5. See error > the modal is not displayed

I attached a screen record

https://user-images.githubusercontent.com/16067358/154268656-310faa1d-45d4-4f6a-8c12-43b213bcb497.mp4


**Issue3**
1. Enable multistore & Create a new shop
2. Go to BO > Shop Parameters > General page
3. See an error in the console 

image
**Without PR** => OK, I don't have any error.

**Issue4**
I don't know if it is a bug or this is the wanted behavior

1. Enable multistore & Create a new shop
2. Go to BO > Shop Parameters > General > Maintenance page
3. Make a change & don't save
4. Click on the header dropdown
5. Click on view my shop
6. The modal "Switch context?" is displayed
7. Click on "Switch context anyway"
8. Redirection to the Home page on the same page & not in opened in a new window 

I attached a screen record

https://user-images.githubusercontent.com/16067358/154273986-f8886e6d-84ac-4b78-b302-cbf8f8f13bd9.mp4

Thanks!

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@sowbiba Could you feedback from @khouloudbelguith  ?

**Comment by matks** [comment]
Hi @sowbiba unfortunately this PR is stale. I close it after 30 days without activity. But you can reopen it if needed 😄

**Pull request closed without merging**
